File: src/rag/generate.ts
import type { Hit } from "@/rag/types";
import { tokenize } from "@/rag/tokenize";

export const ABSTAIN =
  "در منابع بازیابی‌شده مدرک کافی نیست. پاسخ را حدس نمی‌زنم.";

function sentences(text: string): string[] {
  return text
    .split(/(?<=[.!?؟।])\s+|\n+/)
    .map((s) => s.trim())
    .filter((s) => s.length > 20);
}

function overlapScore(queryTokens: string[], sentence: string): number {
  const st = new Set(tokenize(sentence));
  if (st.size === 0 || queryTokens.length === 0) return 0;
  let hit = 0;
  for (const t of queryTokens) if (st.has(t)) hit += 1;
  return hit / queryTokens.length;
}

function chunkOverlap(queryTokens: string[], text: string): number {
  const st = new Set(tokenize(text));
  if (!queryTokens.length) return 0;
  let hit = 0;
  for (const t of queryTokens) if (st.has(t)) hit += 1;
  return hit / queryTokens.length;
}

export function generateGroundedAnswer(query: string, hits: Hit[]): {
  answer: string;
  abstained: boolean;
  citations: string[];
} {
  if (hits.length === 0) {
    return { answer: ABSTAIN, abstained: true, citations: [] };
  }

  const qTokens = tokenize(query);
  if (chunkOverlap(qTokens, hits[0].chunk.text) < 0.08) {
    return { answer: ABSTAIN, abstained: true, citations: [] };
  }

  const scored: { text: string; chunkId: string; score: number }[] = [];
  for (const hit of hits) {
    for (const s of sentences(hit.chunk.text)) {
      scored.push({
        text: s,
        chunkId: hit.chunk.id,
        score: overlapScore(qTokens, s) + 1 / (hit.rank + 4),
      });
    }
  }
  scored.sort((a, b) => b.score - a.score);

  const picked: typeof scored = [];
  const used = new Set<string>();
  for (const item of scored) {
    const key = item.text.slice(0, 80);
    if (used.has(key)) continue;
    used.add(key);
    picked.push(item);
    if (picked.length >= 3) break;
  }

  if (picked.length === 0) {
    const fallback = sentences(hits[0].chunk.text).slice(0, 2);
    if (fallback.length === 0) {
      return { answer: ABSTAIN, abstained: true, citations: [] };
    }
    const citations = [hits[0].chunk.id];
    return {
      answer: fallback.map((s) => `${s} [${hits[0].chunk.id}]`).join(" "),
      abstained: false,
      citations,
    };
  }

  const citations = [...new Set(picked.map((p) => p.chunkId))];
  const answer = picked.map((p) => `${p.text} [${p.chunkId}]`).join(" ");
  return { answer, abstained: false, citations };
}

File: backends/generate.py
"""Extractive generator. Port of src/rag/generate.ts, including its abstain rule."""

from __future__ import annotations

import re

from backends.base import SearchHit
from backends.tokenize import tokenize
from backends.tracing import call_traced

ABSTAIN = "در منابع بازیابی‌شده مدرک کافی نیست. پاسخ را حدس نمی‌زنم."

_SENTENCE = re.compile(r"(?<=[.!?؟।])\s+|\n+")


def _sentences(text: str) -> list[str]:
    return [part.strip() for part in _SENTENCE.split(text) if len(part.strip()) > 20]


def _overlap_count(query_tokens: list[str], text: str) -> int:
    present = set(tokenize(text))
    return sum(1 for token in query_tokens if token in present)


def _overlap_score(query_tokens: list[str], sentence: str) -> float:
    if not query_tokens:
        return 0.0
    return _overlap_count(query_tokens, sentence) / len(query_tokens)


def _generate(query: str, hits: list[SearchHit]) -> dict[str, object]:
    if not hits:
        return {"answer": ABSTAIN, "abstained": True, "citations": []}

    query_tokens = tokenize(query)
    if _overlap_count(query_tokens, hits[0]["text"]) < 2:
        return {"answer": ABSTAIN, "abstained": True, "citations": []}

    scored: list[dict[str, object]] = []
    for rank, hit in enumerate(hits, start=1):
        for sentence in _sentences(hit["text"]):
            overlap = _overlap_count(query_tokens, sentence)
            score = _overlap_score(query_tokens, sentence)
            if overlap < 2 and score < 0.22:
                continue
            scored.append(
                {
                    "text": sentence,
                    "chunkId": hit["chunk_id"],
                    "score": score + 1 / (rank + 8),
                }
            )
    scored.sort(key=lambda item: float(item["score"]), reverse=True)

    picked: list[dict[str, object]] = []
    used: set[str] = set()
    for item in scored:
        key = str(item["text"])[:80]
        if key in used:
            continue
        used.add(key)
        picked.append(item)
        if len(picked) >= 3:
            break

    if not picked:
        return {"answer": ABSTAIN, "abstained": True, "citations": []}

    citations: list[str] = []
    for item in picked:
        chunk_id = str(item["chunkId"])
        if chunk_id not in citations:
            citations.append(chunk_id)
    answer = " ".join(f"{item['text']} [{item['chunkId']}]" for item in picked)
    return {"answer": answer, "abstained": False, "citations": citations}


def generate_grounded_answer(query: str, hits: list[SearchHit]) -> dict[str, object]:
    return call_traced("generate", lambda: _generate(query, hits))
